Feat: Operation, Attempt, and GFE metrics

lszinv opened the pull request:
> Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

## Description

Implemented metric emission logic for the following metrics:

Operation Latency
Operation Count
Attempt Latency
Attempt Count

Metrics are currently defaulted off, and can be enabled by setting the env var `SPANNER_ENABLE_BUILTIN_METRICS` to `true`

## Impact

> What's the impact of this change?
When enabled, operation and attempt metrics will be captured and emitted to GCM. 

## Testing

> Have you added unit and integration tests if necessary?
> Were any tests changed? Are any breaking changes necessary?

## Additional Information

> Any additional details that we should be aware of?

## Checklist

- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-spanner/issues/new/choose) before writing your code! That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease
- [ ] Appropriate docs were updated
- [ ] Appropriate comments were added, particularly in complex areas or places that require background
- [ ] No new warnings or issues will be generated from this change

Fixes #issue_number_goes_here 🦕

lszinv committed:
Added requestId to interceptor fetching of Tracer

lszinv committed:
feat: Operation and Attempt metrics

Added operation and attempt metrics emissions
- Operation Latency
- Operation Count
- Attempt Latency
- Attempt Count
- GFE Latency
- GFE Connectivity Error Count

Metrics are feature gated to the env var
  `SPANNER_DISABLE_BUILTIN_METRICS` being set to `true`

- Metrics are disabled when emulator is used

lszinv commented on src/v1/spanner_client.ts:
```diff
@@ -315,6 +315,14 @@ export class SpannerClient {
+            const metricsFactory = MetricsTracerFactory.getInstance();
+            let project = "", instance = "", database = "";
+            if (metricsFactory.enabled) {
+               ({project, instance, database} = metricsFactory.getInstanceAttributes(args[0]['name'] ?? args[0]['session']?? args[0]['database']));
```
We had previously talked about getting this information from inside their respective classes instead. 

i.e. get the project from the Spanner client, the instance from the instances stored in the client, and the database stored in instances. 

However, if I did that, I would need to keep the tracer objects in the database classes, and would require changing of function signatures just to pass the reference down to this callPromise to access the tracer.

Instead, if I construct the tracer in this call, then i can just make use of the project/instance/database data that is already available here and feels easier to maintain without introducing more direct dependencies.

lszinv commented on src/metrics/metrics-tracer-factory.ts:
```diff
@@ -122,49 +114,11 @@ export class MetricsTracerFactory {
-    return this._clientAttributes;
-  }
-
-  set project(project: string) {
```
Refactored these to have the factory hold the values directly, and only the ones resource attributes.

The metric attributes are set only once the tracer is created and refactored out to not be in the factory.

lszinv commented on src/index.ts:
```diff
@@ -1507,6 +1513,22 @@ class Spanner extends GrpcService {
+   */
+  configureMetrics_() {
+    const metricsEnabled =
+      process.env.SPANNER_ENABLE_BUILTIN_METRICS === 'true';
```
I can go make a PR to fix the Python client once finished with the changes here.

lszinv commented:
Rebased over main. 

system-tests are failing, but i'm seeing the same failures when running the tests on the main branch. Seems connected to the InstanceAdminClient, which shouldn't be affected by metrics integration.

surbhigarg92 commented on test/metrics/metrics.ts:
```diff
@@ -0,0 +1,130 @@
+      // Metric Attempts should be recorded 3 times from the interceptor:
+      // - 1 for the session
+      // - 2 for the retried executeStreamingSql
+      assert.strictEqual(mockMetricsTracer.recordAttemptStart.callCount, 3);
```
We need more mockSpanner test for verifying status and other attributes for different scenarios. Refer https://github.com/googleapis/java-spanner/blob/main/google-cloud-spanner/src/test/java/com/google/cloud/spanner/OpenTelemetryBuiltInMetricsTracerTest.java

lszinv commented on system-test/spanner.ts:
```diff
@@ -230,6 +231,8 @@ describe('Spanner', () => {
 
   before(async () => {
+    MetricsTracerFactory.resetInstance();
+    process.env.SPANNER_DISABLE_BUILTIN_METRICS = 'true';
```
Ah right, sorry, had temporarily left that part out os that it didn't interfere with my testing during development and forgot to add it afterwards.

lszinv commented:
> > Rebased over main.
> > system-tests are failing, but i'm seeing the same failures when running the tests on the main branch. Seems connected to the InstanceAdminClient, which shouldn't be affected by metrics integration.
> 
> The error seems to be related to this PR.
> 
> ```
> 1) Spanner
>        Transactions
>          batch transactions
>            should create and execute a query partition using callback:
>      Uncaught TypeError: formattedName.match is not a function
>       at MetricsTracerFactory.getInstanceAttributes ([src/metrics/metrics-tracer-factory.ts:235](https://cs.corp.google.com/piper///depot/google3/src/metrics/metrics-tracer-factory.ts?l=235):33)
>       at MetricsTracerFactory.createMetricsTracer ([src/metrics/metrics-tracer-factory.ts:202](https://cs.corp.google.com/piper///depot/google3/src/metrics/metrics-tracer-factory.ts?l=202):12)
>       at <anonymous> ([src/batch-transaction.ts:189](https://cs.corp.google.com/piper///depot/google3/src/batch-transaction.ts?l=189):47)
>       at <anonymous> ([src/instrument.ts:177](https://cs.corp.google.com/piper///depot/google3/src/instrument.ts?l=177):16)
>       at AsyncHooksContextManager.with (node_modules/@[opentelemetry/context-async-hooks/src/AsyncHooksContextManager.ts:49](https://cs.corp.google.com/piper///depot/google3/opentelemetry/context-async-hooks/src/AsyncHooksContextManager.ts?l=49):17)
>       at ContextAPI.with (node_modules/@[opentelemetry/api/src/api/context.ts:77](https://cs.corp.google.com/piper///depot/google3/opentelemetry/api/src/api/context.ts?l=77):42)
>       at NoopTracer.startActiveSpan (node_modules/@[opentelemetry/api/src/trace/NoopTracer.ts:98](https://cs.corp.google.com/piper///depot/google3/opentelemetry/api/src/trace/NoopTracer.ts?l=98):27)
>       at ProxyTracer.startActiveSpan (node_modules/@[opentelemetry/api/src/trace/ProxyTracer.ts:51](https://cs.corp.google.com/piper///depot/google3/opentelemetry/api/src/trace/ProxyTracer.ts?l=51):20)
>       at startTrace ([src/instrument.ts:139](https://cs.corp.google.com/piper///depot/google3/src/instrument.ts?l=139):49)
>       at BatchTransaction.createQueryPartitions ([src/batch-transaction.ts:183](https://cs.corp.google.com/piper///depot/google3/src/batch-transaction.ts?l=183):22)
>       at BatchTransaction.wrapper (node_modules/@[google-cloud/promisify/build/src/index.js:33](https://cs.corp.google.com/piper///depot/google3/google-cloud/promisify/build/src/index.js?l=33):35)
>       at <anonymous> (system-[test/spanner](http://test/spanner).ts:9723:24)
>       at <anonymous> ([src/database.ts:942](https://cs.corp.google.com/piper///depot/google3/src/database.ts?l=942):13)
>       at <anonymous> ([src/transaction.ts:514](https://cs.corp.google.com/piper///depot/google3/src/transaction.ts?l=514):13)
>       at <anonymous> ([src/index.ts:1626](https://cs.corp.google.com/piper///depot/google3/src/index.ts?l=1626):15)
>       at Immediate.wrappedCallback ([src/v1/spanner_client.ts:1564](https://cs.corp.google.com/piper///depot/google3/src/v1/spanner_client.ts?l=1564):11)
>       at process.processImmediate (node:[internal/timers:478](https://cs.corp.google.com/piper///depot/google3/internal/timers?l=478):21)
>       at process.callbackTrampoline (node:[internal/async_hooks:128](https://cs.corp.google.com/piper///depot/google3/internal/async_hooks?l=128):17)
> 
>   2) Spanner
>        Transactions
>          batch transactions
>            should create and execute a query partition using await:
>      AssertionError [ERR_ASSERTION]: ifError got unwanted exception: formattedName.match is not a function
>       at Context.<anonymous> (system-[test/spanner](http://test/spanner).ts:9799:18)
>       at MetricsTracerFactory.getInstanceAttributes ([src/metrics/metrics-tracer-factory.ts:235](https://cs.corp.google.com/piper///depot/google3/src/metrics/metrics-tracer-factory.ts?l=235):33)
>       at MetricsTracerFactory.createMetricsTracer ([src/metrics/metrics-tracer-factory.ts:202](https://cs.corp.google.com/piper///depot/google3/src/metrics/metrics-tracer-factory.ts?l=202):12)
>       at <anonymous> ([src/batch-transaction.ts:189](https://cs.corp.google.com/piper///depot/google3/src/batch-transaction.ts?l=189):47)
>       at <anonymous> ([src/instrument.ts:177](https://cs.corp.google.com/piper///depot/google3/src/instrument.ts?l=177):16)
>       at AsyncHooksContextManager.with (node_modules/@[opentelemetry/context-async-hooks/src/AsyncHooksContextManager.ts:49](https://cs.corp.google.com/piper///depot/google3/opentelemetry/context-async-hooks/src/AsyncHooksContextManager.ts?l=49):17)
>       at ContextAPI.with (node_modules/@[opentelemetry/api/src/api/context.ts:77](https://cs.corp.google.com/piper///depot/google3/opentelemetry/api/src/api/context.ts?l=77):42)
>       at NoopTracer.startActiveSpan (node_modules/@[opentelemetry/api/src/trace/NoopTracer.ts:98](https://cs.corp.google.com/piper///depot/google3/opentelemetry/api/src/trace/NoopTracer.ts?l=98):27)
>       at ProxyTracer.startActiveSpan (node_modules/@[opentelemetry/api/src/trace/ProxyTracer.ts:51](https://cs.corp.google.com/piper///depot/google3/opentelemetry/api/src/trace/ProxyTracer.ts?l=51):20)
>       at startTrace ([src/instrument.ts:139](https://cs.corp.google.com/piper///depot/google3/src/instrument.ts?l=139):49)
>       at BatchTransaction.createQueryPartitions ([src/batch-transaction.ts:183](https://cs.corp.google.com/piper///depot/google3/src/batch-transaction.ts?l=183):22)
>       at /workspace/node_modules/@[google-cloud/promisify/build/src/index.js:60](https://cs.corp.google.com/piper///depot/google3/google-cloud/promisify/build/src/index.js?l=60):28
>       at new Promise (<anonymous>)
>       at BatchTransaction.wrapper (node_modules/@[google-cloud/promisify/build/src/index.js:45](https://cs.corp.google.com/piper///depot/google3/google-cloud/promisify/build/src/index.js?l=45):16)
>       at Context.<anonymous> (system-[test/spanner](http://test/spanner).ts:9787:31)
>       ```
> ```

Weird, the failure is giving me a different error and stack trace when running locally. I'll fix this.

lszinv commented:
New PR was pushed to address previously discussed issues:
- GFE metric recording moved so that they could contain the status in their attributes
- Operation recording moved to the index's Request and RequestStream functions
- Additional tests added to verify all the metrics

surbhigarg92 commented on src/metrics/metrics-tracer.ts:
```diff
@@ -27,220 +28,417 @@ import {
+  /**
+   * Sets the client UID attribute if not already set.
+   */
   set clientUid(clientUid: string) {
     if (!(METRIC_LABEL_KEY_CLIENT_UID in this._clientAttributes)) {
       this._clientAttributes[METRIC_LABEL_KEY_CLIENT_UID] = clientUid;
     }
   }
 
+  /**
+   * Sets the client name attribute if not already set.
+   */
   set clientName(clientName: string) {
     if (!(METRIC_LABEL_KEY_CLIENT_NAME in this._clientAttributes)) {
       this._clientAttributes[METRIC_LABEL_KEY_CLIENT_NAME] = clientName;
     }
   }
```
These two client attributes will never change throughout the course of application,  so setting it for every metric tracer created can be an overkill, maybe we can find a way to add these common attributes which are not specific to request either during opentelemetry/view creation or in exporter?

lszinv replied:
This was actually why in a previous commit, they were being included as a part of the Resource attributes. However, that was changed to only hold the resource metrics to keep things logically consistent. 

I'll make changes to hold the value in the Factory, and have the exporter query for it when translating between metric types.

surbhigarg92 commented on test/session-pool.ts:
```diff
@@ -90,12 +104,32 @@ describe('SessionPool', () => {
   };
 
   before(() => {
+    // Set up OpenTelemetry tracer provider for testing
+    const exporter = new InMemorySpanExporter();
+    tracerProvider = new NodeTracerProvider({
+      sampler: new AlwaysOnSampler(),
+      exporter: exporter,
+      spanProcessors: [new SimpleSpanProcessor(exporter)],
+    });
+    tracerProvider.register();
+
+    // Set up context manager
+    contextManager = new AsyncHooksContextManager();
+    contextManager.enable();
+    context.setGlobalContextManager(contextManager);
+
```
Why these changes are required for this PR ?

lszinv replied:
Earlier testing implementations were overlapping with the session-pool tests as they were sharing the implicitly defined global contexts in OTEL. I've since changed it, but kept this code there to guard the tests against similar issues in the future. 

As they are now though, the metric tests are no longer sharing any of the global contexts and won't clash. I'll remove it to keep the change list specific to the features being added.

surbhigarg92 reviewed (commented):
While trying to run this on my local, I am seeing two below errors

1. `(node:2233) MetadataLookupWarning: received unexpected error = request to http://169.254.169.254/computeMetadata/v1/instance failed, reason: connect ENOTSUP 169.254.169.254:80 - Local (192.0.0.2:63313) code = ENOTSUP
(Use `node --trace-warnings ...` to show where the warning was created)`

I am guessing this is happening when we try to fetch `location` attribute. We should handle such error and print a custom log in this case something like, location attribute could not be fetched as application is not on GCP , falling back to "global"


2. `ERROR: Send TimeSeries failed: 3 INVALID_ARGUMENT: One or more TimeSeries could not be written: timeSeries[0-9] (example metric.type="spanner.googleapis.com/internal/client/operation_latencies", metric.labels={"status": "OK", "method": "executeStreamingSql", "client_name": "spanner-nodejs/8.0.0", "client_uid": "c8fdea05-6422-4195-9929-0760a4e401ec@2233@gargsurbhi-macbookpro.roam.internal", "database": "test-db"}): write for resource=spanner_instance_client{location:global,instance_id:unknown,instance_config:unknown,client_hash:00036b} failed with: One or more points were written more frequently than the maximum sampling period configured for the metric. {Metric: spanner.googleapis.com/internal/client/attempt_latencies, Timestamps: {Youngest Existing: '2025/07/02-01:36:52.720', New: '2025/07/02-01:37:06.042'}}`

I am guessing this is happening at the time of application shutdown. when application shutdown happens the metric exporter tries to do an export, and if the export has happened recently , less than 60 sec ago, in this case we see this error. Similar issue was fixed for Go sometime back https://github.com/googleapis/google-cloud-go/pull/12266/files

surbhigarg92 commented:
When I ran the metrics on my machine, I saw a lot of difference in attempt latency and operation latency data for batchCreateSessions , for executeStreamingSql method, it was almost similar
<img width="1476" alt="Screenshot 2025-07-02 at 2 44 29 PM" src="https://github.com/user-attachments/assets/1475af87-cc4f-4718-a253-1accccd82ee0" />

lszinv commented on src/metrics/metrics-tracer-factory.ts:
```diff
@@ -16,169 +16,288 @@ import * as crypto from 'crypto';
+    if (!MetricsTracerFactory.enabled) {
+      return null;
+    }
+    if (this._currentOperationTracers.has(method.toLowerCase())) {
+      return this._currentOperationTracers.get(method.toLowerCase());
+    }
```
this is the problem. I'll have a change pushed out with this fixed once I reproduce @surbhigarg92 's code and verify the fix.

lszinv committed:
Addressed Review Comments

  - Disabled exporting of metrics if already exported within last 30s
  - Added View definitions for bucket boundaries for histograms
  - Fixed issue with CreateTracer reusing previously defined
     operation.
  - Moved Client Name and UID setting to exporter
  - Improved tracer associations with attempts by using google request
    id instead of the method

surbhigarg92 commented on test/metrics/metrics.ts:
```diff
@@ -0,0 +1,407 @@
+        getAggregatedValue(operationCountData, executeAttributes),
+        1,
+      );
+      getAggregatedValue(operationLatenciesData, executeAttributes);
+      assert.strictEqual(
+        getAggregatedValue(attemptCountData, executeAttributes),
+        2,
+      );
```
This is not correct. Currently attemptCount is 2, but both are having OK status. 

It should be 1 attempt with "UNAVAILABLE" status and other with "OK"

surbhigarg92 commented on test/metrics/metrics.ts:
```diff
@@ -0,0 +1,407 @@
+      await database.close();
+    });
+
+    it('should increase attempts on retries', async () => {
```
This only tests retries happening from Spanner client, we should also track retries from GAX client. Currently GAX only retries non streaming calls, use below test . 

```
it.only('should increase attempts on retries for non streaming calls with gax options', async () => {
      gfeStub = sandbox
        .stub(MetricsTracer.prototype, 'extractGfeLatency')
        .callsFake((header: string) => 123);
      const database = newTestDatabase();
      const err = {
        message: 'Temporary unavailable',
        code: grpc.status.UNAVAILABLE,
      } as MockError;
      spannerMock.setExecutionTime(
        spannerMock.commit,
        SimulatedExecutionTime.ofError(err),
      );

      const GAX_OPTIONS = {
        retry: {
          retryCodes: [4, 8, 14],
          backoffSettings: {
            initialRetryDelayMillis: 1000,
            retryDelayMultiplier: 1.3,
            maxRetryDelayMillis: 32000,
            initialRpcTimeoutMillis: 60000,
            rpcTimeoutMultiplier: 1,
            maxRpcTimeoutMillis: 60000,
            totalTimeoutMillis: 600000,
          },
        },
      };
      await database.runTransactionAsync(
              async tx => {
                await tx.run(selectSql);
// Commit RPC will be retried by GAX
                await tx.commit({gaxOptions: GAX_OPTIONS});
              },
            );

      const {resourceMetrics} = await reader.collect();

      const operationCountData = getMetricData(
        resourceMetrics,
        METRIC_NAME_OPERATION_COUNT,
      );
     // Attempt count is correct here but status of attempts are not correct
      const attemptCountData = getMetricData(
        resourceMetrics,
        METRIC_NAME_ATTEMPT_COUNT,
      );
      const operationLatenciesData = getMetricData(
        resourceMetrics,
        METRIC_NAME_OPERATION_LATENCIES,
      );
      const attemptLatenciesData = getMetricData(
        resourceMetrics,
        METRIC_NAME_ATTEMPT_LATENCIES,
      );
      const gfeLatenciesData = getMetricData(
        resourceMetrics,
        METRIC_NAME_GFE_LATENCIES,
      );

      const sessionAttributes = {
        ...commonAttributes,
        database: `database-${dbCounter}`,
        method: 'batchCreateSessions',
      };
      // Verify batchCreateSession metrics are unaffected
      assert.strictEqual(
        getAggregatedValue(operationCountData, sessionAttributes),
        1,
      );
      getAggregatedValue(operationLatenciesData, sessionAttributes);
      assert.strictEqual(
        getAggregatedValue(attemptCountData, sessionAttributes),
        1,
      );
      getAggregatedValue(attemptLatenciesData, sessionAttributes);
      assert.strictEqual(
        getAggregatedValue(gfeLatenciesData, sessionAttributes),
        123,
      );

      const executeAttributes = {
        ...commonAttributes,
        database: `database-${dbCounter}`,
        method: 'executeStreamingSql',
      };
      // Verify executeStreamingSql has 2 attempts and 1 operation
      assert.strictEqual(
        getAggregatedValue(operationCountData, executeAttributes),
        1,
      );
      getAggregatedValue(operationLatenciesData, executeAttributes);
      assert.strictEqual(
        getAggregatedValue(attemptCountData, executeAttributes),
        2,
      );
      getAggregatedValue(attemptLatenciesData, executeAttributes);
      assert.strictEqual(
        getAggregatedValue(gfeLatenciesData, executeAttributes),
        123,
      );
    });
    ```

surbhigarg92 reviewed (commented):
Running unit test on local is throwing error  ERROR: Send TimeSeries failed: 7 PERMISSION_DENIED: Permission monitoring.timeSeries.create denied (or the resource may not exist).
Need monitoring metric writer permission on project observability-project-id. Follow https://cloud.google.com/spanner/docs/view-manage-client-side-metrics#access-client-side-metrics to set up permissions

lszinv commented on src/index.ts:
```diff
@@ -1711,21 +1747,51 @@ class Spanner extends GrpcService {
         if (err) {
           callback(err);
+          metricsTracer?.recordOperationCompletion();
         } else {
-          requestFn(callback);
+          const wrappedCallback = (...args) => {
+            callback(...args);
+            metricsTracer?.recordOperationCompletion();
+          };
+          requestFn(wrappedCallback);
```
moved.

lszinv commented:
> Running unit test on local is throwing error  ERROR: Send TimeSeries failed: 7 PERMISSION_DENIED: Permission monitoring.timeSeries.create denied (or the resource may not exist). Need monitoring metric writer permission on project observability-project-id. Follow https://cloud.google.com/spanner/docs/view-manage-client-side-metrics#access-client-side-metrics to set up permissions

Are these permissions not specific to the auth'd user/service account where the tests are being run?

lszinv commented on src/metrics/metrics-tracer.ts:
```diff
@@ -25,223 +26,336 @@ import {
 
   constructor() {
     this._attemptCount = 0;
     this._startTime = new Date(Date.now());
```
deleted the start() method

lszinv committed:
Addressed review comments

  - Added periodic tracer cleanup to MetricsTracerFactory
  - Changed RequestStream operation completion logic to better handle
    error cases

lszinv committed:
Added tests for tracer pruning

surbhigarg92 committed:
Merge branch 'main' into operation_attempt_metrics

surbhigarg92 commented:
> > Running unit test on local is throwing error  > ERROR: Send TimeSeries failed: 7 PERMISSION_DENIED: Permission monitoring.timeSeries.create denied (or the resource may not exist). Need monitoring metric writer permission on project observability-project-id. Follow https://cloud.google.com/spanner/docs/view-manage-client-side-metrics#access-client-side-metrics to set up permissions
> 
> Are these permissions not specific to the auth'd user/service account where the tests are being run?

But I am talking about unit tests, why running unit tests should require any permissions

lszinv commented on src/metrics/constants.ts:
```diff
@@ -19,7 +19,9 @@ import {
 export const SPANNER_RESOURCE_TYPE = 'spanner_instance_client';
-
+// Maximum time to keep MetricsTracers before considering them stale, and stop tracking them.
+export const TRACER_CLEANUP_THRESHOLD_MS = 10 * 60 * 1000; // 10 minutes
```
Ok will do. I was arbitrarily choosing values, and was hoping you would have better ones :)

lszinv committed:
Addressed cleanup intervals and running logic

lszinv committed:
testing system test cleanup

lszinv committed:
Added better cleanup to tests

- Disabled metrics for unit tests

lszinv committed:
Added Env var setting changes to owlbot to persist between generations

lszinv commented:
> _Warning_: This pull request is touching the following templated files:
> 
>     * .github/workflows/ci.yaml - `.github/workflows/ci.yaml` (GitHub Actions) should be updated in [`synthtool`](https://github.com/googleapis/synthtool)

New commit now has a change to owlboy.py to add the env var

lszinv committed:
disable metrics for unit tests

- Disabled Metrics from unit tests by setting the disable ENV var in
  command definition
- Added MetricsTracerFactory cleanup to spanner.close()

lszinv committed:
reverted package and fixed metrics permission

- Disabled metrics from inside tests instead

lszinv committed:
unref the MetricsTracerFactory Interval
